pkg/apis/messaging/v1alpha1/pipeline_types.go

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2019 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+import (
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	"github.com/knative/pkg/apis"
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	"github.com/knative/pkg/webhook"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type Pipeline struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the Pipeline.
+	Spec PipelineSpec `json:"spec,omitempty"`
+
+	// Status represents the current state of the Pipeline. This data may be out of
+	// date.
+	// +optional
+	Status PipelineStatus `json:"status,omitempty"`
+}
+
+// Check that Pipeline can be validated, can be defaulted, and has immutable fields.
+var _ apis.Validatable = (*Pipeline)(nil)
+var _ apis.Defaultable = (*Pipeline)(nil)
+var _ apis.Immutable = (*Pipeline)(nil)
+var _ runtime.Object = (*Pipeline)(nil)
+var _ webhook.GenericCRD = (*Pipeline)(nil)
+
+type PipelineSpec struct {
+	// Steps is the list of Subscribers (processors / functions) that will be called in the order
+	// provided.
+	Steps []eventingv1alpha1.SubscriberSpec
+
+	// TODO: Specify the CRD Channel which should be used to create the underlying Channels.
+
+	// Subscriber is the addressable that optionally receives replies from the last step in the pipeline.
+	// +optional
+	Subscriber *eventingv1alpha1.SubscriberSpec `json:"subscriber,omitempty"`
```

Comment:
Why is this needed? Can it be the last `Steps`?

Reply:
Good question. I got this type wrong. It should have been ReplyStrategy or something like that (it assumes currently only a channel, but there's an issue to relax this constraint to others as well). Basically for the last subscription (representing the last function in the pipeline) created by the pipeline controller, there needs to be a way to specify the Subscription.Spec.Reply. For example, if I have a pipeline with 2 functions and the last one produces the "final" output of this pipeline, I might want to wire that into the Broker for others to consume, or push it to some other channel if strictly operating in the messaging (Channel / Subscription bases).